pkg/sync/syncer.go

```diff
@@ -1452,6 +1472,532 @@ func (s *syncer) syncGrantsForResource(ctx context.Context, resourceID *v2.Resou
+	}
+
+	err = s.store.PutEntitlements(ctx, ents...)
+	if err != nil {
+		return err
+	}
+
+	err = s.store.PutGrants(ctx, grantsForEnts...)
+	if err != nil {
+		return err
+	}
+
+	l.Info("Synced external resources",
+		zap.Int("resource-types", len(resourceTypes)),
+		zap.Int("resources", len(principals)),
+		zap.Int("entitlements", len(ents)),
+		zap.Int("grants", len(grantsForEnts)),
+	)
+
+	err = s.processGrantsWithExternalPrincipals(ctx, principals)
+	if err != nil {
+		return err
+	}
+
+	s.state.FinishAction(ctx)
+
+	return nil
+}
+
+func (s *syncer) SyncExternalResourcesUsersAndGroups(ctx context.Context) error {
+	ctx, span := tracer.Start(ctx, "syncer.SyncExternalResourcesUsersAndGroups")
+	defer span.End()
+
+	l := ctxzap.Extract(ctx)
+	l.Info("Syncing external resources for users and groups...")
+
+	skipEGForResourceType := make(map[string]bool)
+
+	resourceTypes, err := s.listExternalResourceTypes(ctx)
+	if err != nil {
+		return err
+	}
+
+	userAndGroupResourceTypes := make([]*v2.ResourceType, 0)
+	ents := make([]*v2.Entitlement, 0)
+	principals := make([]*v2.Resource, 0)
+	grantsForEnts := make([]*v2.Grant, 0)
+	for _, rt := range resourceTypes {
+		for _, t := range rt.Traits {
+			if t == v2.ResourceType_TRAIT_USER || t == v2.ResourceType_TRAIT_GROUP {
+				userAndGroupResourceTypes = append(userAndGroupResourceTypes, rt)
+				continue
+			}
+		}
+	}
+
+	for _, rt := range userAndGroupResourceTypes {
+		rtAnnos := annotations.Annotations(rt.Annotations)
+		skipEntitlements := rtAnnos.Contains(&v2.SkipEntitlementsAndGrants{})
+		skipEGForResourceType[rt.Id] = skipEntitlements
+
+		resourceListResp, err := s.listExternalResourcesForResourceType(ctx, rt.Id)
+		if err != nil {
+			return err
+		}
+
+		for _, resourceVal := range resourceListResp {
+			resourceAnnos := annotations.Annotations(resourceVal.GetAnnotations())
+			batonID := &v2.BatonID{}
+			resourceAnnos.Update(batonID)
+			resourceVal.Annotations = resourceAnnos
+			principals = append(principals, resourceVal)
+		}
+	}
+
+	for _, principal := range principals {
+		skipEnts := skipEGForResourceType[principal.Id.ResourceType]
+		if skipEnts {
+			continue
+		}
+		resourceEnts, err := s.listExternalEntitlementsForResource(ctx, principal)
+		if err != nil {
+			return err
+		}
+		ents = append(ents, resourceEnts...)
+	}
+
+	for _, ent := range ents {
+		grantsForEnt, err := s.listExternalGrantsForEntitlement(ctx, ent)
+		if err != nil {
+			return err
+		}
+		grantsForEnts = append(grantsForEnts, grantsForEnt...)
+	}
+
+	err = s.store.PutResourceTypes(ctx, userAndGroupResourceTypes...)
+	if err != nil {
+		return err
+	}
+
+	err = s.store.PutResources(ctx, principals...)
+	if err != nil {
+		return err
+	}
+
+	err = s.store.PutEntitlements(ctx, ents...)
+	if err != nil {
+		return err
+	}
+
+	err = s.store.PutGrants(ctx, grantsForEnts...)
+	if err != nil {
+		return err
+	}
+
+	l.Info("Synced external resources",
+		zap.Int("resource-types", len(userAndGroupResourceTypes)),
```

Comment:
These are counts so it'd be nice if the zap field names were "resource_type_count", "resource_count", etc. The same goes for the l.Info() in SyncExternalResources().